File: Source/Core/Entities/UnhookableTileController.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.Aqua.Core
{
    [CustomEntity("Aqua/Unhookable Tile Controller")]
    public class UnhookableTileController : Entity
    {
        public char BlockTile1 { get; private set; }
        public char BlockTile2 { get; private set; }
        public char BlockTile3 { get; private set; }
        public char BlockTile4 { get; private set; }
        public bool Activate1 { get; private set; }
        public bool Activate2 { get; private set; }
        public bool Activate3 { get; private set; }
        public bool Activate4 { get; private set; }

        public UnhookableTileController(EntityData data, Vector2 offset)
            : base()
        {
            BlockTile1 = data.Char("block_tile1");
            BlockTile2 = data.Char("block_tile2");
            BlockTile3 = data.Char("block_tile3");
            BlockTile4 = data.Char("block_tile4");
            Activate1 = data.Bool("activate1");
            Activate2 = data.Bool("activate2");
            Activate3 = data.Bool("activate3");
            Activate4 = data.Bool("activate4");
        }

        public override void Update()
        {
            base.Update();
            var state = SceneAs<Level>().GetState();
            if (state != null)
            {
                if (Activate1 && BlockTile1 != '\0')
                    state.UnhookableTiletypes.Add(BlockTile1);
                if (Activate2 && BlockTile2 != '\0')
                    state.UnhookableTiletypes.Add(BlockTile2);
                if (Activate3 && BlockTile3 != '\0')
                    state.UnhookableTiletypes.Add(BlockTile3);
                if (Activate4 && BlockTile4 != '\0')
                    state.UnhookableTiletypes.Add(BlockTile4);
            }
            RemoveSelf();
        }
    }
}

File: Loenn/entities/unhookable_tile_controller.lua
local DrawableRectangle = require("structs.drawable_rectangle")
local DrawableText = require("structs.drawable_text")
local Depths = require("consts.object_depths")
local FakeTilesHelper = require("helpers.fake_tiles")

local UnhookableTileController = {}

UnhookableTileController.name = "Aqua/Unhookable Tile Controller"
UnhookableTileController.fieldInformation = {}
UnhookableTileController.fieldInformation = FakeTilesHelper.addTileFieldInformation(UnhookableTileController.fieldInformation, "block_tile1")
UnhookableTileController.fieldInformation = FakeTilesHelper.addTileFieldInformation(UnhookableTileController.fieldInformation, "block_tile2")
UnhookableTileController.fieldInformation = FakeTilesHelper.addTileFieldInformation(UnhookableTileController.fieldInformation, "block_tile3")
UnhookableTileController.fieldInformation = FakeTilesHelper.addTileFieldInformation(UnhookableTileController.fieldInformation, "block_tile4")
UnhookableTileController.placements = {
    name = "Unhookable Tile Controller",
    data = {
        block_tile1 = FakeTilesHelper.getPlacementMaterial(),
        block_tile2 = FakeTilesHelper.getPlacementMaterial(),
        block_tile3 = FakeTilesHelper.getPlacementMaterial(),
        block_tile4 = FakeTilesHelper.getPlacementMaterial(),
        activate1 = true,
        activate2 = false,
        activate3 = false,
        activate4 = false,
        clear = false,
    },
}

function UnhookableTileController.sprite(room, entity)
    local x = entity.x or 0
    local y = entity.y or 0

    local width = entity.width or 16
    local height = entity.height or 16

    local fillColor, borderColor, textColor = "0cfefe88", "ffffff", "ffffff"
    local borderedRectangle = DrawableRectangle.fromRectangle("bordered", x, y, width, height, fillColor, borderColor)
    -- local textDrawable = DrawableText.fromText("Unhookable Tile Controller", x, y, width, height, nil, Triggers.triggerFontSize, textColor)

    local drawables = borderedRectangle:getDrawableSprite()
    for _, drawable in ipairs(drawables) do
        drawable.depth = Depths.triggers
    end
    -- table.insert(drawables, textDrawable)

    -- textDrawable.depth = Depths.triggers - 1

    return drawables
end

return UnhookableTileController